File: leetcode/20e.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for t in s:
            if t in ['(', '[', '{']:
                stack.insert(0, t)
                continue
            if not stack:
                return False  # no matching bracket
            if t == ')':
                top = stack.pop(0)
                if top != '(':
                    return False
            elif t == ']':
                top = stack.pop(0)
                if top != '[':
                    return False
            elif t == '}':
                top = stack.pop(0)
                if top != '{':
                    return False
        return not stack

File: leetcode/20e.java
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (isOpen(c)) {
                stack.add(c);
            } else if (isClose(c)) {
                // can't peek unless i check if it's empty first, are you fucking serious
                if (stack.isEmpty()) {
                    return false;
                }
                char d = stack.peek();
                if (getComplement(d) == c) {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }

    public boolean isOpen(char c) {
        return c == '(' || c == '{' || c == '[';
    }

    public boolean isClose(char c) {
        return c == ')' || c == '}' || c == ']';
    }

    public char getComplement(char c) {
        if (c == '(') {
            return ')';
        } else if (c == '[') {
            return ']';
        } else if (c == '{') {
            return '}';
        }
        return '_';
    }
}
